File: homeAssistant/src/session/ContextRetriever.ts
import type { Database } from 'better-sqlite3';
import type { ContextSpec, ContextResult, FilterParams } from '../types';
import { EmbeddingService } from './EmbeddingService';
import { CONTEXT_CONFIG } from '../config/context';

const TABLE_META: Record<string, {
    textCol: string;
    hasUserId: boolean;
    dateCol?: string;
    vecTable?: string;
}> = {
    memos:          { textCol: 'content',     hasUserId: true,  dateCol: 'created_at', vecTable: 'vec_memos'   },
    todos:          { textCol: 'content',     hasUserId: true,  dateCol: 'created_at', vecTable: 'vec_todos'   },
    schedules:      { textCol: 'title',       hasUserId: true,  dateCol: 'event_date'                         },
    home_status:    { textCol: 'device_name', hasUserId: false, dateCol: 'updated_at'                         },
    item_locations: { textCol: 'item_name',   hasUserId: false, dateCol: 'updated_at'                         },
    assets:         { textCol: 'category',    hasUserId: true,  dateCol: 'recorded_at'                        },
    recipes:        { textCol: 'name',        hasUserId: true,  dateCol: 'created_at', vecTable: 'vec_recipes' },
    grocery_items:  { textCol: 'name',        hasUserId: false                                                },
};

const ALLOWED_TABLES = new Set(Object.keys(TABLE_META));

export class ContextRetriever {
    constructor(
        private readonly db: Database,
        private readonly embedding: EmbeddingService,
    ) {}

    async retrieve(specs: ContextSpec[], userId: string): Promise<ContextResult[]> {
        return Promise.all(specs.map(spec => this.retrieveOne(spec, userId)));
    }

    async storeEmbedding(table: string, vecTable: string, rowId: number, text: string): Promise<void> {
        const embedding = await this.embedding.embed(text);
        this.embedding.store(vecTable, rowId, embedding);
    }

    private async retrieveOne(spec: ContextSpec, userId: string): Promise<ContextResult> {
        switch (spec.type) {
            case 'recent':  return this.retrieveRecent(spec.db, userId);
            case 'query':   return this.retrieveQuery(spec.db, userId, spec.filter ?? {});
            case 'similar': return this.retrieveSimilar(spec.db, userId, spec.searchText ?? '');
        }
    }

    private retrieveRecent(table: string, userId: string): ContextResult {
        if (!ALLOWED_TABLES.has(table)) return { db: table, type: 'recent', rows: [] };
        const meta = TABLE_META[table]!;
        const dateCol = meta.dateCol ?? 'rowid';
        const rows = meta.hasUserId
            ? this.db.prepare(
                `SELECT * FROM ${table} WHERE (user_id = ? OR is_shared = 1) ORDER BY ${dateCol} DESC LIMIT ?`
              ).all(userId, CONTEXT_CONFIG.recentLimit) as Record<string, unknown>[]
            : this.db.prepare(
                `SELECT * FROM ${table} ORDER BY ${dateCol} DESC LIMIT ?`
              ).all(CONTEXT_CONFIG.recentLimit) as Record<string, unknown>[];
        return { db: table, type: 'recent', rows };
    }

    private retrieveQuery(table: string, userId: string, filter: FilterParams): ContextResult {
        if (!ALLOWED_TABLES.has(table)) return { db: table, type: 'query', rows: [] };
        const meta = TABLE_META[table]!;
        const conditions: string[] = [];
        const params: unknown[] = [];

        if (meta.hasUserId) {
            conditions.push('(user_id = ? OR is_shared = 1)');
            params.push(userId);
        }
        if (filter.keyword) {
            conditions.push(`${meta.textCol} LIKE ?`);
            params.push(`%${filter.keyword}%`);
        }
        if (filter.dateFrom && meta.dateCol) {
            conditions.push(`date(${meta.dateCol}) >= ?`);
            params.push(filter.dateFrom);
        }
        if (filter.dateTo && meta.dateCol) {
            conditions.push(`date(${meta.dateCol}) <= ?`);
            params.push(filter.dateTo);
        }
        if (filter.category) {
            conditions.push('category = ?');
            params.push(filter.category);
        }
        if (filter.isShared !== undefined) {
            conditions.push('is_shared = ?');
            params.push(filter.isShared ? 1 : 0);
        }

        const where = conditions.length ? `WHERE ${conditions.join(' AND ')}` : '';
        const rows = this.db.prepare(
            `SELECT * FROM ${table} ${where} LIMIT ?`
        ).all(...params, CONTEXT_CONFIG.recentLimit) as Record<string, unknown>[];
        return { db: table, type: 'query', rows };
    }

    private async retrieveSimilar(table: string, userId: string, searchText: string): Promise<ContextResult> {
        if (!ALLOWED_TABLES.has(table)) return { db: table, type: 'similar', rows: [] };
        const meta = TABLE_META[table]!;
        if (!meta.vecTable) return this.retrieveRecent(table, userId);

        const queryEmbedding = await this.embedding.embed(searchText);
        const similar = this.embedding.findSimilar(meta.vecTable, queryEmbedding, CONTEXT_CONFIG.similarLimit);
        if (!similar.length) return { db: table, type: 'similar', rows: [] };

        const ids = similar.map(s => Number(s.rowid));
        const placeholders = ids.map(() => '?').join(',');
        const rows = this.db.prepare(
            `SELECT * FROM ${table} WHERE id IN (${placeholders})`
        ).all(...ids) as Record<string, unknown>[];
        return { db: table, type: 'similar', rows };
    }
}

File: src/main/kotlin/com/homeassistant/context/ContextRetriever.kt
package com.homeassistant.context

import com.homeassistant.models.ContextResult
import com.homeassistant.models.ContextSpec
import com.homeassistant.models.FilterParams
import org.jetbrains.exposed.sql.transactions.TransactionManager
import org.jetbrains.exposed.sql.transactions.transaction
import org.slf4j.LoggerFactory

private val log = LoggerFactory.getLogger(ContextRetriever::class.java)

private const val RECENT_LIMIT = 10
private const val SIMILAR_LIMIT = 5

private data class TableMeta(
    val textCol: String,
    val hasUserId: Boolean,
    val hasCategory: Boolean = false,
    val dateCol: String? = null,
    val vecTable: String? = null,
)

private val TABLE_META: Map<String, TableMeta> = mapOf(
    "memos"          to TableMeta("content",     true,  dateCol = "created_at", vecTable = "vec_memos"),
    "todos"          to TableMeta("content",     true,  dateCol = "created_at", vecTable = "vec_todos"),
    "schedules"      to TableMeta("title",       true,  dateCol = "event_date"),
    "home_status"    to TableMeta("device_name", false, dateCol = "updated_at"),
    "item_locations" to TableMeta("item_name",   false, dateCol = "updated_at"),
    "assets"         to TableMeta("category",    true,  hasCategory = true, dateCol = "recorded_at"),
    "recipes"        to TableMeta("name",        true,  dateCol = "created_at", vecTable = "vec_recipes"),
    "grocery_items"  to TableMeta("name",        false),
)

private val ALLOWED_TABLES = TABLE_META.keys.toSet()

class ContextRetriever(private val embedding: EmbeddingService) {

    fun retrieve(specs: List<ContextSpec>, userId: String): List<ContextResult> =
        specs.map { retrieveOne(it, userId) }

    fun storeEmbedding(table: String, rowId: Long, text: String) {
        val meta = TABLE_META[table] ?: return
        val vecTable = meta.vecTable ?: return
        val vector = embedding.embed(text) ?: return
        try {
            val conn = TransactionManager.current().connection.connection as java.sql.Connection
            embedding.store(conn, vecTable, rowId, vector)
        } catch (e: Exception) {
            log.warn("storeEmbedding failed for $table/$rowId: ${e.message}")
        }
    }

    private fun retrieveOne(spec: ContextSpec, userId: String): ContextResult =
        when (spec.type) {
            "recent"  -> retrieveRecent(spec.db, userId)
            "query"   -> retrieveQuery(spec.db, userId, spec.filter ?: FilterParams())
            "similar" -> retrieveSimilar(spec.db, userId, spec.searchText ?: "")
            else      -> ContextResult(spec.db, spec.type, emptyList())
        }

    private fun retrieveRecent(table: String, userId: String): ContextResult {
        if (table !in ALLOWED_TABLES) return ContextResult(table, "recent", emptyList())
        val meta = TABLE_META[table]!!
        val dateCol = meta.dateCol ?: "rowid"
        val rows = transaction {
            val sql = if (meta.hasUserId)
                "SELECT * FROM $table WHERE (user_id = ? OR is_shared = 1) ORDER BY $dateCol DESC LIMIT ?"
            else
                "SELECT * FROM $table ORDER BY $dateCol DESC LIMIT ?"
            execRaw(sql, if (meta.hasUserId) listOf(userId, RECENT_LIMIT) else listOf(RECENT_LIMIT))
        }
        return ContextResult(table, "recent", rows)
    }

    private fun retrieveQuery(table: String, userId: String, filter: FilterParams): ContextResult {
        if (table !in ALLOWED_TABLES) return ContextResult(table, "query", emptyList())
        val meta = TABLE_META[table]!!
        val conditions = mutableListOf<String>()
        val params = mutableListOf<Any>()

        if (meta.hasUserId) {
            conditions.add("(user_id = ? OR is_shared = 1)")
            params.add(userId)
        }
        filter.keyword?.let {
            conditions.add("${meta.textCol} LIKE ?")
            params.add("%$it%")
        }
        if (filter.dateFrom != null && meta.dateCol != null) {
            conditions.add("date(${meta.dateCol}) >= ?")
            params.add(filter.dateFrom)
        }
        if (filter.dateTo != null && meta.dateCol != null) {
            conditions.add("date(${meta.dateCol}) <= ?")
            params.add(filter.dateTo)
        }
        if (filter.category != null && meta.hasCategory) {
            conditions.add("category = ?")
            params.add(filter.category)
        }
        if (filter.isShared != null && !meta.hasUserId) {
            conditions.add("is_shared = ?")
            params.add(if (filter.isShared) 1 else 0)
        }

        val where = if (conditions.isEmpty()) "" else "WHERE ${conditions.joinToString(" AND ")}"
        val rows = transaction {
            execRaw("SELECT * FROM $table $where LIMIT ?", params + RECENT_LIMIT)
        }
        return ContextResult(table, "query", rows)
    }

    private fun retrieveSimilar(table: String, userId: String, searchText: String): ContextResult {
        if (table !in ALLOWED_TABLES) return ContextResult(table, "similar", emptyList())
        val meta = TABLE_META[table]!!
        if (meta.vecTable == null) return retrieveRecent(table, userId)

        val queryEmbedding = embedding.embed(searchText) ?: return retrieveRecent(table, userId)

        val similar = transaction {
            val conn = TransactionManager.current().connection.connection as java.sql.Connection
            embedding.findSimilar(conn, meta.vecTable, queryEmbedding, SIMILAR_LIMIT)
        }
        if (similar.isEmpty()) return ContextResult(table, "similar", emptyList())

        val ids = similar.map { it.rowid }
        val placeholders = ids.joinToString(",") { "?" }
        val rows = transaction {
            execRaw("SELECT * FROM $table WHERE id IN ($placeholders)", ids)
        }
        return ContextResult(table, "similar", rows)
    }

    /** Execute a raw SQL SELECT and return results as list of maps. */
    private fun execRaw(sql: String, params: List<Any>): List<Map<String, Any?>> {
        val conn = TransactionManager.current().connection.connection as java.sql.Connection
        val results = mutableListOf<Map<String, Any?>>()
        conn.prepareStatement(sql).use { stmt ->
            params.forEachIndexed { i, v ->
                when (v) {
                    is Int    -> stmt.setInt(i + 1, v)
                    is Long   -> stmt.setLong(i + 1, v)
                    is String -> stmt.setString(i + 1, v)
                    else      -> stmt.setObject(i + 1, v)
                }
            }
            val rs = stmt.executeQuery()
            val meta = rs.metaData
            while (rs.next()) {
                val row = mutableMapOf<String, Any?>()
                for (col in 1..meta.columnCount) {
                    row[meta.getColumnName(col)] = rs.getObject(col)
                }
                results.add(row)
            }
        }
        return results
    }
}
